src/cpp/session/modules/rmarkdown/NotebookAlternateEngines.cpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * NotebookAlternateEngines.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+
+#include "SessionRmdNotebook.hpp"
+#include "SessionExecuteChunkOperation.hpp"
+#include "NotebookCache.hpp"
+#include "NotebookAlternateEngines.hpp"
+
+#include <boost/algorithm/string.hpp>
+
+#include <core/StringUtils.hpp>
+#include <core/Algorithm.hpp>
+#include <core/Exec.hpp>
+#include <core/json/JsonRpc.hpp>
+
+#include <r/RExec.hpp>
+
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+namespace {
+
+class ChunkExecCompletedScope
+{
+public:
+   ChunkExecCompletedScope(const std::string& docId,
+                           const std::string& chunkId)
+      : docId_(docId), chunkId_(chunkId)
+   {
+   }
+   
+   ~ChunkExecCompletedScope()
+   {
+      events().onChunkExecCompleted(
+               docId_,
+               chunkId_,
+               notebookCtxId());
+   }
+   
+private:
+   std::string docId_;
+   std::string chunkId_;
+};
+
+Error prepareCacheConsoleOutputFile(const std::string& docId,
+                                    const std::string& chunkId,
+                                    const std::string& nbCtxId,
+                                    FilePath* pChunkOutputFile)
+{
+   // forward declare error
+   Error error;
+   
+   // prepare chunk directory
+   FilePath cachePath = notebook::chunkOutputPath(
+            docId,
+            chunkId,
+            notebook::ContextExact);
+   
+   error = cachePath.removeIfExists();
```

Comment:
You can call `FilePath::resetDirectory` to get the combo of removeIfExists/ensureDirectory